packages/ai/core/generate-text/generate-text.ts

```diff
@@ -469,9 +469,35 @@ A function that attempts to repair a tool call that failed to parse.
             ? text + stepText
             : stepText;
 
-        currentReasoningDetails = asReasoningDetails(
-          currentModelResponse.reasoning,
-        );
+        // Only add reasoning details if they exist and aren't already included
+        if (
+          currentModelResponse.reasoning &&
+          currentModelResponse.reasoning.length > 0
+        ) {
+          const newReasoningDetails = asReasoningDetails(
+            currentModelResponse.reasoning,
+          );
+          // Avoid duplicate reasoning by checking if we already have this content
+          const reasoningExists = currentReasoningDetails.some(
+            existing =>
+              JSON.stringify(existing) === JSON.stringify(newReasoningDetails),
+          );
+
+          if (!reasoningExists) {
+            // Handle both cases: single ReasoningDetail or array of ReasoningDetail
+            if (Array.isArray(newReasoningDetails)) {
+              currentReasoningDetails = [
+                ...currentReasoningDetails,
+                ...newReasoningDetails,
+              ];
+            } else {
+              currentReasoningDetails = [
+                ...currentReasoningDetails,
+                newReasoningDetails,
+              ];
+            }
+          }
+        }
```

Comment:
Why is this necessary? my instinct says this should be solved in the provider. Is the bedrock provider returning duplicated reasoning?

Reply:
I have that here because let's say you pass a tool to the aisdk with an `execute()` method. That means you'll get two responses from the sdk. The first response will have the currentReasoningDetails, but that isn't extracted out to the generateText response from the provider, because the SECOND run (after the tool result is done), will overwrite the reasoning details. This code just appends the reasoning details every time this is called for that interaction, ensuring we always get the reasoning

Reply:
Okay, do you want me to extract the reasoning from steps in the model provider, or would you like for me to leave as is and wait for an upgrade `ai` package?

Reply:
leave as is and wait for `ai` package upgrade. the extraction would be in the user code

Reply:
I wonder if this was related to #4986...